### ISSUE TITLE
Factory workarounds

### ISSUE THREAD

**Issue by egeerardyn** [matlab2tikz/matlab2tikz#441]
Since #377, we have noticed that some Factory settings in MATLAB do not map well to the (effective) standard settings for some objects.
This is not a major bug on our side, we would just generate code that is too verbose and wouldn't use TikZ defaults in those cases (just a minor annoyance).
We have to work-around this issue, in particular for the following values:
- R2014a (and older?): 
  - `FactoryAxesXColor` (and probably the Y and Z variants as well) are set to white, while the effective value is black.
- R2014b (and newer?):
  - `FactoryLegendEdgeColor` has value 20% gray, but the effective value is 15% gray

In short, we have to hard-code the effective values somewhere in the code. Unfortunately, those are dependent on the MATLAB version.

**Comment by PeterPablo** [comment]
Just a short remark here: In my opinion it is not necessary to explicitly store some default values, since we can always compare the property (e.g., color of legend's edge) to its factory value `FactoryLegendEdgeColor`. If they are identical, we omit it. If they differ, we proceed as usual.

**Comment by egeerardyn** [comment]
@PeterPablo I agree. But we have to take care of that edge case we discovered in #641: MATLAB sets `DefaultAxesColorOrder` by itself as well in HG1, so we cannot rely on `FactoryAxesColorOrder` there.

**Comment by PeterPablo** [comment]
Yes. I'd say, we have a closer look at this, once #641 is merged.

**Comment by CattusNiger** [comment]
Maybe what I'm going to say is a huge piece of shit (that's the good thing of having no idea, I can yell a lot of crap and nothing happens), but if in the end we all want to have the same output in tikz, no matter what Matlab or Octave version we have, wouldn't it be better to compare the properties of the plot with the default properties in the tikz library instead of the default in Matlab/Octave?

I don't know if this is possible/realistic/easy, but it was the first thing that come to my head once I reach this problem.

Hope this could help, or at least don't offend.

**Comment by PeterPablo** [comment]
Thank you for sharing your thoughts. Unfortunately matlab and octave do
behave differently. We thus need to check against their default properties
and only if there is a discrepancy pass those properties on to pgfplots.

**Comment by egeerardyn** [comment]
@CattusNiger Thanks for joining the discussion. What you say, certainly is valuable and I think it is indeed feasible by hardcoding the pgfplots defaults into the program (quite some work, though). Typically they are quite stable. But do note, as @PeterPablo said: different versions of MATLAB/Octave behave differently. E.g. in R2014b+ the axes are somewhat grayish in color, while they are black in Octave and R2014a-. So the output will differ for those, unless you specifically set the color to e.g. red.

Let me first state that I have mixed feelings about all possible approaches here: checking the MATLAB defaults or checking pgfplots defaults since in all cases there is some use case where this is not the desired/consistent behavior.

Personally, I prefer our current approach since this _tries_ to interpret the figure, e.g. a plain MATLAB call `axis()`  should translate to a plain pgfplots `\begin{axis} \end{axis}`. Using `set(gca, ...)` the corresponding options are appended. `grid on` will add `grid=both` or something similar to the output. I admit, however, that this _interpretation_ is absolutely not perfect, but I think it's the lesser evil.

The other approach (i.e. `'strict', true`) we currently have is to just translate whatever we find. In that case `grid on` (which produces dotted grids in MATLAB) will also produce dotted grids in pgfplots (unlike the former which just produces the default pgfplots grid).

In a way, I see your proposal as a halfway option between both: keep the output code minimal (i.e. when the value set in MATLAB corresponds with the default in pgfplots), but strictly replicate how the figure looks (in standard pgfplots, so when one does not use `\pgfplotsset{every axis/.style={...}}` commands).

But your argument makes me wonder whether we couldn't make a loadable list of options and corresponding values that are translated to an empty string in pgfplots. E.g. such a list for "strict" mode would be empty, our current approach would contain all values from `get(0,'factory')` (and some patches as depicted in this thread) and what you propose would contain the pgfplots defaults translated into MATLAB terms.

**Comment by CattusNiger** [comment]
But, if I understand it well, the following situation may be possible:
- One property X is not modified in a plot and, thus, has de default value in Matlab/Octave.
- For this reason, the Tikz/PGF library draws it with its default property.
- In this case, the property's default in Matlab/Octave is different from Tikz/PGF and the plots look different under Matlab/Octave and LaTeX.

Is this correct? Maybe there is no possibility, but at least it seems feasible.

Thanks for your answers, by the way.

**Comment by egeerardyn** [comment]
@CattusNiger That's indeed correct (and I'm quite positive there are some examples of that behavior).
Using strict mode (or what you proposed), the same should not happen.

**Comment by CattusNiger** [comment]
Yes, you're right. But I was thinking in adding the less redundant code in the LaTeX file. But if you have it in mind, despite it may be not possible to add it, I'm not worried.
